liblocation/src/main/java/com/mapbox/android/core/location/LocationCallbackTransport.java

```diff
@@ -0,0 +1,118 @@
+package com.mapbox.android.core.location;
+
+import android.location.Location;
+import android.location.LocationListener;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
+import android.os.Message;
+import android.util.Log;
+
+import java.lang.ref.WeakReference;
+
+import static com.mapbox.android.core.location.Utils.isBetterLocation;
+
+/**
+ * A base wrap class for LocationListener, contains the logic of downsample
+ * location update frequent.
+ */
+public class LocationCallbackTransport implements LocationListener {
+  private static final String TAG = "CallbackTransport";
+  private static final int MESSAGE_UPDATE_LOCATION = 0;
+  private static final int MESSAGE_NOT_CACHE_PERIOD = 1;
+  protected final LocationEngineCallback<LocationEngineResult> callback;
+  private Location currentBestLocation;
+  private long fastestInterval = 0;
+  private Location cacheLocation;
+  private Handler handler;
+
+  LocationCallbackTransport(LocationEngineCallback<LocationEngineResult> callback) {
+    this.callback = callback;
+  }
+
+  private static class ThresholdHandler extends Handler {
+    private final WeakReference<LocationCallbackTransport> weakReference;
+
+    ThresholdHandler(Looper looper, LocationCallbackTransport callbackTransport) {
+      super(looper);
+      weakReference = new WeakReference<>(callbackTransport);
```

Comment:
under which conditions we're expecting that strong reference to callback may cause a leak?